dll/win32/kernel32/client/console/console.c

```diff
@@ -3030,13 +3030,115 @@ SetConsoleLocalEUDC(DWORD Unknown1, DWORD Unknown2, DWORD Unknown3, DWORD Unknow
     return FALSE;
 }
 
+static BOOL
+IntRegisterConsoleIME(
+    HWND hWnd,
+    DWORD dwThreadId,
+    DWORD cbDesktop,
+    LPWSTR lpDesktop,
+    LPDWORD pdwAttachTo)
+{
+    BOOL ret;
+    CONSOLE_API_MESSAGE ApiMessage;
+    PCONSOLE_REGISTERCONSOLEIME RegisterConsoleIME = &ApiMessage.Data.RegisterConsoleIME;
+    PCSR_CAPTURE_BUFFER pCaptureBuffer;
+    DWORD dwError;
+
+    RegisterConsoleIME->ConsoleHandle = NtCurrentPeb()->ProcessParameters->ConsoleHandle;
+    RegisterConsoleIME->hWnd = hWnd;
+    RegisterConsoleIME->dwThreadId = dwThreadId;
+    RegisterConsoleIME->cbDesktop = cbDesktop;
+
+    pCaptureBuffer = CsrAllocateCaptureBuffer(1, cbDesktop);
+    if (!pCaptureBuffer)
+    {
+        SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+        return FALSE;
+    }
+
+    CsrCaptureMessageBuffer(pCaptureBuffer,
+                            lpDesktop,
+                            cbDesktop,
+                            (PVOID*)&RegisterConsoleIME->pDesktop);
+    CsrClientCallServer((PCSR_API_MESSAGE)&ApiMessage,
+                        pCaptureBuffer,
+                        CSR_CREATE_API_NUMBER(CONSRV_SERVERDLL_INDEX, ConsolepRegisterConsoleIME),
+                        sizeof(*RegisterConsoleIME));
+    CsrFreeCaptureBuffer(pCaptureBuffer);
+
+    if (!NT_SUCCESS(ApiMessage.Status))
+    {
+        dwError = RtlNtStatusToDosError(ApiMessage.Status);
+        SetLastError(dwError);
+        return FALSE;
+    }
+
+    ret = TRUE;
+    if (pdwAttachTo)
+    {
+        __TRY
```

Comment:
Where did you get these names? Please use SEH2.

Reply:
Fixed.